File: src/io/fdeitylink/keroedit/image/PxAttr.java
package io.fdeitylink.keroedit.image;

import java.text.MessageFormat;

import java.nio.file.Files;
import java.nio.file.Path;

import java.nio.file.StandardOpenOption;

import java.nio.channels.SeekableByteChannel;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import java.io.IOException;
import java.text.ParseException;

import io.fdeitylink.util.NullArgumentException;

import io.fdeitylink.keroedit.Messages;

//TODO: Move back to PxAttrManager as static inner class?
public final class PxAttr {
    private static final String HEADER_STRING = "pxMAP01\0";

    private final Path path;

    private int[][] attributes;

    PxAttr(final Path inPath) throws IOException, ParseException {
        path = NullArgumentException.Companion.requireNonNull(inPath, "PxAttr", "inPath").toAbsolutePath();

        /*if (!Files.exists(inPath)) {
            inPath = Paths.get(inPath.getParent().toAbsolutePath().toString() + File.separatorChar + "mpt00" + GameData.pxAttrExtension);
            if (!Files.exists(inPath)) {
                throw new FileNotFoundException(Messages.INSTANCE.getString("PxAttrManager.PxAttr.DEFAULT_MISSING"));
            }
        }*/

        try (SeekableByteChannel chan = Files.newByteChannel(inPath, StandardOpenOption.READ)) {
            ByteBuffer buf = ByteBuffer.allocate(HEADER_STRING.length());
            chan.read(buf);

            if (!(new String(buf.array()).equals(HEADER_STRING))) {
                throw new ParseException(MessageFormat.format(Messages.INSTANCE.getString("PxAttrManager.PxAttr.INCORRECT_HEADER"),
                                                              inPath.getFileName()),
                                         (int)chan.position());
            }

            buf = ByteBuffer.allocate(4);
            buf.order(ByteOrder.LITTLE_ENDIAN);
            chan.read(buf);
            buf.flip();

            //TODO: validate dimensions/size
            final int width = buf.getShort();
            final int height = buf.getShort();

            if (width * height > 0) {
                //TODO: Find if it is ever not 0 (might've already checked but make sure)
                chan.position(chan.position() + 1);

                buf = ByteBuffer.allocate(width * height);
                chan.read(buf);
                buf.flip();

                attributes = new int[height][width];
                for (int y = 0; y < height; ++y) {
                    for (int x = 0; x < width; ++x) {
                        attributes[y][x] = buf.get() & 0xFF; //& 0xFF treats it as unsigned byte when converted to int
                    }
                }
            }
            else {
                attributes = null;
            }
        }
        catch (final IOException except) {
            throw new IOException(MessageFormat.format(Messages.INSTANCE.getString("PxAttrManager.PxAttr.IOEXCEPT"),
                                                       inPath.getFileName()), except);
        }
    }

    //for cloning into new PxAttr
    PxAttr(final PxAttr pxAttr, final Path inPath) {
        attributes = NullArgumentException.Companion.requireNonNull(pxAttr, "PxAttr", "pxAttr").getAttributes();
        path = NullArgumentException.Companion.requireNonNull(inPath, "PxAttr", "inPath").toAbsolutePath();
    }

    public int[][] getAttributes() {
        if (null != attributes) {
            final int[][] attributesCopy = new int[attributes.length][attributes[0].length];
            for (int y = 0; y < attributes.length; ++y) {
                System.arraycopy(attributes[y], 0, attributesCopy[y], 0, attributes[y].length);
            }
            return attributesCopy;
        }
        return null;
    }

    void setAttribute(final int x, final int y, final int attribute) {
        //TODO: Check attribute for validity (range?)
        attributes[y][x] = attribute;
    }

    void save() throws IOException {
        try (SeekableByteChannel chan = Files.newByteChannel(path,
                                                             StandardOpenOption.WRITE,
                                                             StandardOpenOption.TRUNCATE_EXISTING,
                                                             StandardOpenOption.CREATE)) {
            ByteBuffer buf = ByteBuffer.wrap(HEADER_STRING.getBytes("SJIS"));
            chan.write(buf);

            if (null == attributes) {
                buf = ByteBuffer.allocate(4);
                buf.putShort((short)0).putShort((short)0);
                buf.flip();

                chan.write(buf);
            }
            else {
                short width = (short)attributes[0].length;
                short height = (short)attributes.length;

                buf = ByteBuffer.allocate(5);
                buf.order(ByteOrder.LITTLE_ENDIAN);
                buf.putShort(width).putShort(height);
                buf.put((byte)0);
                buf.flip();

                chan.write(buf);

                buf = ByteBuffer.allocate((width & 0xFFFF) * (height & 0xFFFF)); //& 0xFFFF treats as unsigned when converted to int
                for (final int[] row : attributes) {
                    for (final int attr : row) {
                        buf.put((byte)attr);
                    }
                }
                buf.flip();
                chan.write(buf);
            }
        }
    }
}

File: src/io/fdeitylink/keroedit/image/PxAttr.kt
package io.fdeitylink.keroedit.image

import java.text.MessageFormat

import java.nio.file.Path
import java.nio.file.Files

import java.nio.file.StandardOpenOption

import java.nio.charset.Charset

import java.nio.ByteBuffer
import java.nio.ByteOrder

import java.text.ParseException
import java.io.IOException

import kotlin.experimental.and

import io.fdeitylink.util.Array2D

import io.fdeitylink.util.Logger

import io.fdeitylink.keroedit.Messages

class PxAttr {
    companion object {
        //TODO: Create default Array2D<Int> for attributes with all empty values
        private const val HEADER_STRING = "pxMAP01\u0000"
    }

    private val path: Path

    /*
     * TODO:
     * Rather than making this nullable, use an Array2D<Int> with width and height of 0
     *  - Or with width and height of 16 and all attributes set to 0?
     * Somehow make this a val (is a primary constructor needed for this?)
     */
    private var _attributes: Array2D<Int>? = null
    val attributes: Array2D<Int>?
        get() {
            val attrs = _attributes
            return if (null == attrs) null else Array2D(attrs.width, attrs.height) { x, y -> attrs[x, y] }
        }

    @Suppress("UsePropertyAccessSyntax")
    constructor(inPath: Path) {
        path = inPath.toAbsolutePath()

        if (!Files.exists(path)) {
            //TODO: Rather than setting to null, save a new PxAttr file with empty attributes by calling save()
            _attributes = null
            return
        }

        try {
            Files.newByteChannel(path, StandardOpenOption.READ).use {
                var buf = ByteBuffer.allocate(HEADER_STRING.length)
                it.read(buf)

                if (String(buf.array()) != HEADER_STRING) {
                    throw ParseException(MessageFormat.format(Messages["PxAttr.INCORRECT_HEADER"], path),
                                         it.position().toInt())
                }

                buf = ByteBuffer.allocate(4)
                buf.order(ByteOrder.LITTLE_ENDIAN)
                it.read(buf)
                buf.flip()

                //TODO: Validate dimensions are 16 by 16 or 0 by 0
                val width = buf.getShort().toInt()
                val height = buf.getShort().toInt()

                _attributes = if (width * height > 0) {
                    //TODO: Find if the skipped value is ever not 0 (might've already checked but make sure)
                    it.position(it.position() + 1)

                    buf = ByteBuffer.allocate(width * height)
                    it.read(buf)

                    //TODO: Use buf.array().forEachIndexed()
                    val bufArray = buf.array()
                    Array2D(width, height) { x, y -> (bufArray[(width * y) + x] and 0xFF.toByte()).toInt() }
                }
                else {
                    null
                }
            }
        }
        catch (except: IOException) {
            //TODO: Have PxAttrManager log and throw the specialized exception?
            Logger.logThrowable("Exception while parsing PXATTR file $path", except)
            throw IOException("${except.localizedMessage}\n$path", except)
        }
    }

    constructor(pxAttr: PxAttr, inPath: Path) {
        _attributes = pxAttr.attributes //Copies, they do not reference same object
        path = inPath.toAbsolutePath()
    }

    fun setAttribute(x: Int, y: Int, attribute: Int) {
        //TODO: Check attribute for validity (range)
        if (null == _attributes) {
            throw IndexOutOfBoundsException("attributes is empty")
        }
        _attributes!![x, y] = attribute
    }

    fun save() {
        try {
            Files.newByteChannel(path,
                                 StandardOpenOption.WRITE,
                                 StandardOpenOption.TRUNCATE_EXISTING,
                                 StandardOpenOption.CREATE).use {
                var buf = ByteBuffer.wrap(HEADER_STRING.toByteArray(Charset.forName("SJIS")))
                it.write(buf)

                val attrs = _attributes
                if (null == attrs) {
                    buf = ByteBuffer.allocate(4)
                    buf.putShort(0.toShort())
                            .putShort(0.toShort())
                    buf.flip()

                    it.write(buf)
                    return
                }
                else {
                    val width = attrs.width.toShort()
                    val height = attrs.height.toShort()

                    buf = ByteBuffer.allocate(5)
                    buf.order(ByteOrder.LITTLE_ENDIAN)

                    buf.putShort(width).putShort(height).put(0.toByte())
                    buf.flip()

                    it.write(buf)

                    buf = ByteBuffer.allocate((width and 0xFFFF.toShort()) * (height and 0xFFFF.toShort()))
                    attrs.forEach { buf.put(it.toByte()) }
                    buf.flip()
                    it.write(buf)
                }
            }
        }
        catch(except: IOException) {
            //TODO: Have PxAttrManager log and throw the specialized exception?
            Logger.logThrowable("Exception while saving PXATTR file $path", except)
            throw IOException("${except.localizedMessage}\n$path", except)
        }
    }
}